feat: Language support for mapping programs to logs

jherbage opened the pull request:
Hi @jaqx0r 
This is the parser side of the log mapping.
It joins the other PR at the runtimes code Object with the addition of the slice of logfiles mapped.

I can add an integration test but if I understand you right it won't work until both PRs are there so I presume you are going to want me to add a 3rd PR with the integration test in and some update to the readme.

I am not entirely happy with this PR because my understanding of the lexer is extremely slight. For now users must have a new line at the end of a logmapping statement and the logfiles need to be enclosed in quotes ie they are lexed as STRING. If you forget the quotes they are seen as ID and parsing fails. Similarly if you make logmapping the last time in your program and don't end with a new lijne then parsing would fail. I couldn't find any way to avoid these 2 situations whilst still keep the update working. If you know how to make it work so a new line isnt needed after the logmapping statement I would appreciate it. Also you can pass in multple logmapping statements and they just append together. They would actually be deduplicated before they are used in the map in the runtime.

Issue: #188 and google/mtail#35

jaqx0r commented:
> I can add an integration test but if I understand you right it won't work until both PRs are there so I presume you are going to want me to add a 3rd PR with the integration test in and some update to the readme.

That's a good idea.  I had actually expected that you'd branch off the last PR to make this one, but that was presumptuous of me.

jaqx0r commented on internal/runtime/code/object.go:
```diff
@@ -11,8 +11,9 @@ import (
+	Strings      []string          // Static strings.
+	Regexps      []*regexp.Regexp  // Static regular expressions.
+	Metrics      []*metrics.Metric // Metrics accessible to this program.
+	RelevantLogs []string          // logs this program can be used against - if empty can be applied to any log
```
I think this should be LogRestriction, we haven't yet turned it into a filter, but we've got a list of log names we want to restrict this program to.

jaqx0r commented on internal/runtime/compiler/parser/parser.y:
```diff
@@ -73,6 +77,7 @@ import (
 %token LCURLY RCURLY LPAREN RPAREN LSQUARE RSQUARE
 %token COMMA
 %token NL
+%token LOGMAPPING
```
I think I don't like the keyword name `logmapping`.

Can you try calling it `log_filter`, `logname_filter`, or `filename_filter` or something like that?  I think the correct verb for a user is to filter the log filename, not to map a log to a program, despite the bug's history.

I think `log_filter` is the most descriptive as well as not too verbose.

jherbage replied:
Changed to match suggestion

jaqx0r commented on internal/runtime/compiler/parser/parser.y:
```diff
@@ -142,6 +149,27 @@ stmt
+  {
+    $$ = []string{$1}
+  }
+  | log_mapping_list COMMA STRING
```
Turn this into a list parse like `metric_by_expr` where it has a terminating `id_or_string` and a recursing `log_filter_list, COMMA id_or_string`.

jaqx0r commented on internal/runtime/compiler/parser/parser.y:
```diff
@@ -112,7 +117,9 @@ stmt_list
 /* Types of statements. */
 stmt
-  : conditional_stmt
+  : log_mapping_spec %prec LOGMAPPING
```
For consistency make this a `log_filter_stmt` and then create the intermediate `log_fitler_stmt` where you statr the log_mapping_spec on line 152.

jaqx0r commented on internal/runtime/compiler/parser/parser.y:
```diff
@@ -142,6 +149,27 @@ stmt
   }
   ;
 
+log_mapping_spec
```
Then this should just be

```
log_filter_stmt
 : LOG_FILTER log_filter_list
 ;
```

It's weird to me that you need a NL here unlike the other stmt rules.

jherbage replied:
Done

jherbage commented:
In addition to adding your suggestions @jaqx0r  I added a first attempt at updating the programming docs and removed the TODO entry which sounds its describing what we are adding here.

jherbage commented on internal/runtime/code/object.go:
```diff
@@ -11,8 +11,9 @@ import (
+	Strings        []string          // Static strings.
+	Regexps        []*regexp.Regexp  // Static regular expressions.
+	Metrics        []*metrics.Metric // Metrics accessible to this program.
+	LogRestriction []string          // logs this program can be used against - if empty can be applied to any log
```
Done in PR 214

jherbage commented on docs/Programming-Guide.md:
```diff
@@ -132,6 +132,16 @@ parsing:
+## Log Filtering
+
+The `log_filter` keyword can be used in an mtail program to provide a list of filenames (including the full path to the file). If a `log_filter` is used the program is restricted only to processing logfiles with the names provided. If no `log_filter` keyword is used then all logfiles read by mtail are read by the program. `log_filter` can be used to reduce the resource consumption of mtail by ens ...
+
```
Added

jherbage commented on internal/runtime/compiler/ast/ast.go:
```diff
@@ -296,6 +296,20 @@ func (n *PatternLit) Type() types.Type {
+}
+
+func (n *LogFilter) Type() types.Type {
+	return types.Pattern
```
I tried to switch to use ast's StringLit but the problem with not having its own node type is in the unparsing. The log_filter keyword would not get re-added because using a generic string list wouldn't know it was a log_filter would it? But I could be missing something here as I am still very green on this parser

jherbage commented on internal/runtime/compiler/types/types.go:
```diff
@@ -243,6 +243,8 @@ var (
 	String        = &Operator{"String", []Type{}}
 	Pattern       = &Operator{"Pattern", []Type{}}
+	// LogFilter represents a type for the LOGFILTER construct.
+	LogFilter = &Operator{"LogFilter", []Type{String}}
```
As above. I couldn't get unparsing to work if it wasn't its own ast type

jherbage commented:
I plan to add the integration test in here before final submission of this PR. But it will need 214 to be merged to main and then this rebased off that so a full integration test can work.

jaqx0r commented:
Thanks for the contribution!
